File: app/src/main/java/com/example/aiassistant/Observer.kt
package com.example.aiassistant

import android.content.Context
import android.view.accessibility.AccessibilityEvent

/**
 * Observes user-initiated actions and records them as patterns.
 * This runs in both OBSERVE and ASSIST modes — always learning.
 */
object Observer {

    /**
     * Process a user-initiated action event.
     * Call this when TYPE_VIEW_CLICKED, TYPE_VIEW_SCROLLED, or TYPE_VIEW_TEXT_CHANGED
     * is detected — these represent what the *user* actually did.
     */
    suspend fun onUserAction(
        context: Context,
        event: AccessibilityEvent,
        currentState: String,
        packageName: String
    ) {
        val actionText = extractActionText(event) ?: return
        val actionType = mapEventType(event.eventType)

        DatabaseHelper.recordPattern(
            context = context,
            state = currentState,
            packageName = packageName,
            actionText = actionText,
            actionType = actionType
        )
    }

    private fun extractActionText(event: AccessibilityEvent): String? {
        // Try to get the text of the element the user interacted with
        event.text?.let { texts ->
            val joined = texts.joinToString(" ").trim()
            if (joined.isNotEmpty()) return joined
        }

        event.contentDescription?.toString()?.let {
            if (it.isNotEmpty()) return it
        }

        return null
    }

    private fun mapEventType(eventType: Int): String {
        return when (eventType) {
            AccessibilityEvent.TYPE_VIEW_CLICKED -> "CLICK"
            AccessibilityEvent.TYPE_VIEW_SCROLLED -> "SCROLL"
            AccessibilityEvent.TYPE_VIEW_TEXT_CHANGED -> "TYPE"
            else -> "CLICK"
        }
    }
}

File: BehavioralAssistant-iOS/BehavioralAssistant/Engine/Observer.swift
import Foundation

/// Observes user-initiated actions and records them as patterns.
/// This runs in both OBSERVE and ASSIST modes — always learning.
///
/// On Android, this receives events from the AccessibilityService.
/// On iOS, this must be called explicitly from within the app's own
/// UI event handlers, since cross-app observation is not possible.
enum Observer {

    /// Record a user action.
    ///
    /// - Parameters:
    ///   - actionText: The text/label of the element the user interacted with.
    ///   - eventType: A string describing the interaction ("CLICK", "SCROLL", "TYPE").
    ///   - currentState: The stable structural fingerprint of the current screen.
    ///   - packageName: The bundle identifier of the app being used.
    static func onUserAction(
        actionText: String,
        eventType: String,
        currentState: String,
        packageName: String
    ) {
        PersistenceController.shared.recordPattern(
            state: currentState,
            packageName: packageName,
            actionText: actionText,
            actionType: eventType
        )
    }
}
